core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala

```diff
@@ -475,6 +533,194 @@ object ConsumerGroupCommand extends Logging {
       new KafkaConsumer(properties)
     }
 
+    def resetOffsets(): Map[PartitionAssignmentState, Long] = {
+      val groupId = opts.options.valueOf(opts.groupOpt)
+      val (state, assignments) = collectGroupAssignment(groupId)
+      assignments match {
+        case None =>
+          // applies to both old and new consumer
+          printError(s"The consumer group '$groupId' does not exist.")
+          Map.empty
+        case Some(assignments) =>
+          state match {
+            case Some("Dead") =>
+              printError(s"Consumer group '$groupId' does not exist.")
+              Map.empty
+            case Some("Empty") =>
+              val assignmentsToReset = getAssignmentsToReset(assignments)
+              val assignmentsPrepared = prepareAssignmentsToReset(assignmentsToReset)
+              val execute = opts.options.has(opts.executeOpt)
+              if(execute)
+                resetAssignments(assignmentsPrepared)
+              assignmentsPrepared
+            case Some("PreparingRebalance") | Some("AwaitingSync") =>
+              printError(s"Consumer group '$groupId' offsets cannot be reset if it is rebalancing.")
+              Map.empty
+            case Some("Stable") =>
+              printError(s"Consumer group '$groupId' offsets cannot be reset if it has members active.")
+              Map.empty
+            case other =>
+              // the control should never reach here
+              throw new KafkaException(s"Expected a valid consumer group state, but found '${other.getOrElse("NONE")}'.")
+          }
+      }
+    }
+
+    def getAssignmentsToReset(assignments: Seq[PartitionAssignmentState]): Seq[PartitionAssignmentState] = {
+      if (opts.options.has(opts.allTopicsOpt)) {
+        assignments
+      } else if(opts.options.has(opts.topicOpt)) {
+        val topics = opts.options.valuesOf(opts.topicOpt).asScala
+        val topicsPartitions: Map[String, List[Long]] = topics.map {
```

Comment:
I'd consider converting this to a method. Something like `getPartitionsToReset`.